Finish precompiled functions

mhanberg opened the pull request:
This will precompile the user's images into functions that let them use images from the default collection, custom collections, and insert dynamic HTML attributes.

This also cleans up the documentation and changes the tone to imply that the precompiled functions are the default way to use this library, instead of the "new" way.

Eventually the "old" way should be deprecated and removed from the library.

Addresses #9 

@nikkomiu I would review and merge #13 before reviewing this. I forgot to use a branch for that PR, so until that one is merged, this PR will show both sets of changes. This PR is also dependent on changes from PR #13

mhanberg committed:
Finish implementing pre-compiled svg_image functions

This will precompile the user's images into functions that let them
use images from the default collection, custom collections, and insert
dynamic HTML attributes.

This also cleans up the documentation and changes the tone to imply that
the precompiled functions are the default way to use this library, instead
of the "new" way.

Eventually the "old" way should be deprecated and removed from the
library.

nikkomiu commented:
I agree that the "old" way should probably be removed at this point since the default 😄 way is a much better approach in general.

Sorry, I squashed #13 before I noticed your message about the branch.

mhanberg commented:
No problem. I'll attempt to reconcile my fork and get this mergeable

mhanberg commented:
@nikkomiu This PR should be reviewable now.

nikkomiu commented on lib/phoenix_inline_svg/helpers.ex:
```diff
@@ -264,17 +219,34 @@ defmodule PhoenixInlineSvg.Helpers do
+  defp create_cached_svg_image({collection, name}) do
+    filename =
+      name
+      |> String.split("/")
```
Does it make sense to do this path split and only take the last part of the path? If I have two sub-folders with the same SVG name in it only the last one would be accessible.

Whereas if this wasn't there I would still be able to do:

```elixir
svg_image("some/icon.svg")
# and
svg_image("other/icon.svg")
```

mhanberg replied:
Good catch. I was not thinking of this case since this package has the "collections" concept.

This would currently work with case of `svg_image("icon",  "some")`.

I personally like the API for `svg_image("some/icon")` instead of the "collections" API and would move to deprecate "collections", but would call that out of scope for this PR.

I'll see if I can get this working for now.

nikkomiu commented on lib/phoenix_inline_svg/helpers.ex:
```diff
@@ -264,17 +219,34 @@ defmodule PhoenixInlineSvg.Helpers do
+    svg = read_svg_from_path(name)
 
     quote do
+      def svg_image(unquote(filename)) do
```
Should this function be defined in the macro? I would think this would just be a standard function in the module that uses the default collection to then call the [macro def that follows this one](https://github.com/nikkomiu/phoenix_inline_svg/pull/14/files?short_path=04c6e90#diff-e9df72ca9ba92ba50626afc237069177R236) like the [def on 102](https://github.com/nikkomiu/phoenix_inline_svg/blob/9e496c4ce1620cd500bb1bfaaf5b75ec723e3334/lib/phoenix_inline_svg/helpers.ex#L103):

```elixir
config_or_default(:default_collection, "generic")
```

mhanberg replied:
If this isn't defined in the macro, it doesn't cache the svg as a function. 

Also, the standard functions need to take a `@conn` or an `endpoint`

nikkomiu commented on lib/phoenix_inline_svg/helpers.ex:
```diff
@@ -264,17 +219,34 @@ defmodule PhoenixInlineSvg.Helpers do
+        unquote(PhoenixInlineSvg.Utils.safety_string(svg))
+      end
+
+      def svg_image(unquote(filename), opts) when is_list(opts) do
```
The next def will call the same functions the same way regardless of if opts is a list. This def seems to be redundant because of that (or the other one shouldn't be there).

mhanberg replied:
The functions themselves are both necessary. But I am able to have the lower 1 and 2 arity functions delegate to `svg_image/3`

mhanberg commented:
@nikkomiu Thanks for the review, I should be able to address your comments sometime later in the weekend.

nikkomiu commented:
Awesome, thank you for all of the great work you've done!

mhanberg committed:
Don't generate generic functions for custom collections

mhanberg committed:
Allow for images to be nested in directories inside a collection

mhanberg commented:
@nikkomiu I think I have resolved all of your comments. Let me know if you have any more thoughts.

nikkomiu commented on lib/phoenix_inline_svg/utils.ex:
```diff
@@ -13,4 +13,7 @@ defmodule PhoenixInlineSvg.Utils do
     {:safe, html}
   end
+
+  def insert_generic_funcs(ast, "generic"), do: ast
```
Will need to change this from literal of generic because there is a config setting for overriding the default collection name.

mhanberg replied:
Good, catch! Thanks for the thorough reviews 👍

mhanberg replied:
I pushed a fix for this. Since this is compile time configuration, I don't think there is a way to test it.

mhanberg committed:
Compile functions for default collections

This was previously incorrect due to hard coding "generic" as the
default collection.

mhanberg committed:
Fix multi-word attrs

This was lost during a rebase

nikkomiu commented:
Thanks for all the work you're doing. It's really great!

mhanberg commented:
@nikkomiu no problem!

It would probably be good to cut a release now.
